elasticsearch/file/src/main/java/nl/knaw/huc/service/FieldsService.java

```diff
@@ -0,0 +1,203 @@
+    file.setMetadata(metadata);
+  }
+
+  private void addTypeResource(ResultType type) {
+    var typeJson = getRestResource(TYPE_ENDPOINT, type.getId());
+    type.setName(read(typeJson, "$.name"));
+    type.setMimetype(read(typeJson, "$.mimetype"));
+  }
+
+  private void addDocResource(ResultDoc doc) {
+    var docJson = getRestResource(DOC_ENDPOINT, doc.getId());
+    doc.setExternalId(read(docJson, "$.externalId"));
+  }
+
+  private void addDocMetadataResource(UUID docId, ResultDoc doc) {
+    var docMetadataJson = getRestResource(DOC_METADATA_ENDPOINT, docId);
+    Map<String, String> form = read(docMetadataJson, "$");
+    var metadata = form
+        .entrySet()
+        .stream()
+        .map((entry) -> new ResultMetadataEntry(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+    doc.setMetadata(metadata);
+  }
+
+  private void addVersionsResource(UUID fileId, ResultFields fields) {
+    var versions = new ArrayList<ResultVersion>();
+    fields.setVersions(versions);
+    getAllVersions(fileId, versions);
+    setContentsModified(versions);
+    fields.setContentsLastModified(createContentsLastModified(fields));
+  }
+
+  private void getAllVersions(UUID fileId, ArrayList<ResultVersion> versions) {
+    var versionsUrl = createUrl(textrepoHost, fileId, VERSIONS_ENDPOINT);
+    var ref = new TypeRef<List<FormVersion>>() {};
+
+    var pageTurner = new PageTurner<>(versionsUrl, 0, pageSize, jsonPath, ref);
+    pageTurner.turn(formVersions -> {
+      formVersions.forEach((form) -> versions.add(createResultVersion(form)));
+    });
+  }
+
+  private void setContentsModified(ArrayList<ResultVersion> versions) {
+    for (var index = 0; index < versions.size(); index++) {
+      versions
+          .get(index)
+          .setContentsModified(contentsModified(versions, index));
+    }
+  }
+
+  private ResultVersion createResultVersion(FormVersion form) {
+    var result = new ResultVersion();
+    result.setId(form.getId());
+    result.setSha(form.getSha());
+    result.setCreatedAt(form.getCreatedAt());
+    return result;
+  }
+
+  /**
+   * Determine if contents sha of current version has changed with regard to previous version
+   * Assume all versions are sorted from newest to oldest
+   */
+  private boolean contentsModified(List<ResultVersion> allVersions, int currentIndex) {
+    var isFirstVersion = allVersions.size() == currentIndex + 1;
+    if (isFirstVersion) {
+      return true;
+    }
+    var previous = allVersions.get(currentIndex + 1);
+    var current = allVersions.get(currentIndex);
+    return !current.getSha().equals(previous.getSha());
+  }
+
+  private ResultContentsLastModified createContentsLastModified(ResultFields fields) {
+    var firstVersionWithLatestContents = fields
+        .getVersions()
+        .stream()
+        .filter(ResultVersion::getContentsModified)
+        .findFirst();
+
+    var result = new ResultContentsLastModified();
```

Comment:
Aangezien ResultContentsLastModified geen constructor heeft die zijn velden initialiseert, heeft result hier dus velden in zich die allemaal 'null' zijn. Slechts wanneer firstVersionWithLatestContents 'present' is, worden de velden gevuld, anders retourneer je verderop dus een result met allemaal 'null' velden in zich. Ik overzie even niet of dit de bedoeling is, maar het voelt spannend.

Reply:
Ja, dit was de bedoeling: ik wil een leeg json object teruggeven als er eerste versie (`firstVersionWithLatestContents`) gevonden is.

Reply:
Ok, en 'leeg' is dan null en niet ""

Reply:
leeg is dan `{}`